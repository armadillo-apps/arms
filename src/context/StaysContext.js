import React, { createContext, useContext, useReducer } from "react";
import { staysReducer } from "../reducer/staysReducer";

const StaysContext = createContext({ stays: [] });
export default StaysContext;
export const useStaysContext = () => useContext(StaysContext);

export const StaysProvider = (stays, children) => {
  const [state, dispatch] = useReducer(staysReducer, stays);
  return (
    <StaysContext.Provider value={{ state, dispatch }}>
      {children}
    </StaysContext.Provider>
  );
};
