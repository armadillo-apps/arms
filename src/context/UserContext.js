import React, { createContext, useContext, useReducer } from "react";
import PropTypes from "prop-types";
import { userReducer } from "../reducer/userReducer";

const initialState = {
  email: "",
  role: "",
  loading: false,
  error: false,
  isAuthenticated: false
};

const UserContext = createContext(initialState);
export default UserContext;
export const useUserContext = () => useContext(UserContext);

export const UserProvider = ({ user, children }) => {
  const [state, dispatch] = useReducer(userReducer, user);
  return (
    <UserContext.Provider value={{ state, dispatch }}>
      {children}
    </UserContext.Provider>
  );
};

UserProvider.propTypes = {
  children: PropTypes.element.isRequired,
  user: PropTypes.object.isRequired
};
