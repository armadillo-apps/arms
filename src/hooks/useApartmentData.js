import { useCallback, useEffect, useState } from "react";
import { fetchApartmentById } from "../api/api";

export const useApartmentData = apartmentId => {
  const [apartment, setApartment] = useState({});
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(false);

  const fetchApartment = useCallback(async () => {
    setIsLoading(true);
    try {
      const result = await fetchApartmentById(apartmentId);
      setApartment(result);
    } catch (e) {
      setError(true);
    } finally {
      setIsLoading(false);
    }
  }, [apartmentId]);

  useEffect(() => {
    fetchApartment();
  }, [fetchApartment]);

  return { apartment, isLoading, error, fetchApartment };
};
