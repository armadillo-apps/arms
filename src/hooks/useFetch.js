import { useCallback, useEffect, useState } from "react";

export const useFetch = api => {
  const [data, setData] = useState({});
  const [isFetching, setIsFetching] = useState(false);
  const [isError, setError] = useState(false);

  const fetchData = useCallback(async () => {
    setIsFetching(true);
    try {
      const result = await api();
      setData(result);
    } catch (e) {
      setError(true);
    } finally {
      setIsFetching(false);
    }
  }, [api]);

  useEffect(() => {
    fetchData();
  }, [fetchData]);

  return { data, isFetching, isError, fetchData };
};
