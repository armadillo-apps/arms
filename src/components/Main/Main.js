import React from "react";
import { Redirect, Route, Switch } from "react-router-dom";
import Sidebar from "../../containers/Sidebar";
import routes from "../../router/RouterPaths";
import styles from "./Main.module.scss";
import { useUserContext } from "../../context/UserContext";
import ApartmentProfile from "../../containers/ApartmentProfile";
import OccupantProfile from "../../containers/OccupantProfile";
import { Apartment } from "../Apartment/Apartment";
import Occupants from "../../containers/Occupants";
import NewOccupantForm from "../NewOccupantForm/NewOccupantForm";
import NewApartmentForm from "../NewApartmentForm/NewApartmentForm";
import NewUserForm from "../NewUserForm/NewUserForm";
import ChangePasswordForm from "../ChangePasswordForm/ChangePasswordForm";
import UserManagement from "../UserManagement/UserManagement";
import { roles } from "../../constants/roles";

const RestrictedRoute = ({ component: Component, allowedRoles, ...rest }) => {
  const { state: user } = useUserContext();

  return (
    <Route
      {...rest}
      render={props =>
        user?.isAuthenticated && allowedRoles.includes(user.role) ? (
          <Component {...props} />
        ) : (
          <NoMatchPage />
        )
      }
    />
  );
};

const NoMatchPage = () => {
  return <h1>Path does not exist!</h1>;
};

const Main = () => {
  return (
    <section className={styles.app}>
      <Sidebar />
      <Switch>
        <Route exact path={routes.APARTMENTS} component={Apartment} />
        <Route exact path={routes.MAIN}>
          <Redirect to={routes.APARTMENTS} />
        </Route>
        <Route exact path={routes.OCCUPANTS} component={Occupants} />
        <Route
          path={`${routes.APARTMENTS}/:apartmentId`}
          component={ApartmentProfile}
        />
        <Route
          path={`${routes.OCCUPANTS}/:occupantId`}
          component={OccupantProfile}
        />
        <Route
          exact
          path={routes.CHANGE_PASSWORD}
          component={ChangePasswordForm}
        />
        <RestrictedRoute
          exact
          path={routes.NEW_APARTMENT}
          allowedRoles={[roles.ADMIN, roles.MANAGER]}
          component={NewApartmentForm}
        />
        <RestrictedRoute
          exact
          path={routes.NEW_OCCUPANT}
          allowedRoles={[roles.ADMIN, roles.MANAGER]}
          component={NewOccupantForm}
        />
        <RestrictedRoute
          exact
          path={routes.USERS}
          allowedRoles={[roles.ADMIN]}
          component={UserManagement}
        />
        <RestrictedRoute
          exact
          path={routes.NEW_USER}
          allowedRoles={[roles.ADMIN]}
          component={NewUserForm}
        />
        <Route component={NoMatchPage} />
      </Switch>
    </section>
  );
};

export default Main;
