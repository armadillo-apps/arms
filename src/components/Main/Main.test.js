import React from "react";
import { MemoryRouter, Router } from "react-router-dom";
import { createBrowserHistory } from "history";
import { render } from "@testing-library/react";
import Main from "./Main";
import * as UserContext from "../../context/UserContext";
import routes from "../../router/RouterPaths";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";
import { useFetch } from "../../hooks/useFetch";

jest.mock("../../hooks/useFetch");
jest.mock("react-toast-notifications", () => ({
  useToasts: () => jest.fn()
}));

function renderMain(user, history) {
  jest.spyOn(UserContext, "useUserContext").mockImplementation(() => ({
    state: user,
    dispatch: jest.fn()
  }));

  const utils = render(
    <Router history={history}>
      <Main />
    </Router>
  );
  return { ...utils };
}

describe("Main", () => {
  const history = createBrowserHistory();

  const guestUser = {
    isAuthenticated: true,
    role: "guest"
  };

  const managerUser = {
    isAuthenticated: true,
    role: "manager"
  };

  const adminUser = {
    isAuthenticated: true,
    role: "admin"
  };

  beforeEach(() => {
    jest.clearAllMocks();
    useFetch.mockReturnValue({ data: {} });
  });

  it("renders correctly", () => {
    jest.spyOn(UserContext, "useUserContext").mockImplementation(() => ({
      state: guestUser,
      dispatch: jest.fn()
    }));

    const tree = renderer
      .create(
        <MemoryRouter initialEntries={["/"]}>
          <Main />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("Guest User Navigation", () => {
    it("should redirect authenticated user to apartments page", () => {
      const apartmentUrl = routes.APARTMENTS;
      history.push(apartmentUrl);

      const { getByText } = renderMain(guestUser, history);

      expect(history.location.pathname).toBe(apartmentUrl);
      expect(getByText(/apartment name/i)).toBeInTheDocument();
    });

    it("should redirect authenticated user to occupants page", () => {
      const occupantsUrl = routes.OCCUPANTS;
      history.push(occupantsUrl);

      const { getByText } = renderMain(guestUser, history);

      expect(history.location.pathname).toBe(occupantsUrl);
      expect(getByText(/employee id/i)).toBeInTheDocument();
    });

    it("should redirect authenticated user to change password page", () => {
      const changePasswordUrl = routes.CHANGE_PASSWORD;
      history.push(changePasswordUrl);

      const { getByText } = renderMain(guestUser, history);

      expect(history.location.pathname).toBe(changePasswordUrl);
      expect(getByText(/existing password/i)).toBeInTheDocument();
    });

    it("should not redirect guest user to new apartment page", () => {
      const newApartmentUrl = routes.NEW_APARTMENT;
      history.push(newApartmentUrl);

      const { getByText } = renderMain(guestUser, history);

      expect(history.location.pathname).toBe(newApartmentUrl);
      expect(getByText(/path does not exist!/i)).toBeInTheDocument();
    });

    it("should not redirect guest user to new occupant page", () => {
      const newOccupantUrl = routes.NEW_OCCUPANT;
      history.push(newOccupantUrl);

      const { getByText } = renderMain(guestUser, history);

      expect(history.location.pathname).toBe(newOccupantUrl);
      expect(getByText(/path does not exist!/i)).toBeInTheDocument();
    });

    it("should not redirect guest to new user page", async () => {
      const newUserUrl = routes.NEW_USER;
      history.push(newUserUrl);

      const { getByText } = renderMain(guestUser, history);

      expect(history.location.pathname).toBe(newUserUrl);
      expect(getByText(/path does not exist!/i)).toBeInTheDocument();
    });

    it("should not redirect guest to user management page", async () => {
      const userUrl = routes.USERS;
      history.push(userUrl);

      const { getByText } = renderMain(guestUser, history);

      expect(history.location.pathname).toBe(userUrl);
      expect(getByText(/path does not exist!/i)).toBeInTheDocument();
    });
  });

  describe("Manager Navigation", () => {
    it("should redirect manager to new apartment page", () => {
      const newApartmentUrl = routes.NEW_APARTMENT;
      history.push(newApartmentUrl);

      const { getByText } = renderMain(managerUser, history);

      expect(history.location.pathname).toBe(newApartmentUrl);
      expect(getByText(/create new apartment/i)).toBeInTheDocument();
    });

    it("should redirect manager to new occupant page", async () => {
      const newOccupantUrl = routes.NEW_OCCUPANT;
      history.push(newOccupantUrl);

      const { getByText } = renderMain(managerUser, history);

      expect(history.location.pathname).toBe(newOccupantUrl);
      expect(getByText(/create new occupant/i)).toBeInTheDocument();
    });

    it("should not redirect manager to new user page", async () => {
      const newUserUrl = routes.NEW_USER;
      history.push(newUserUrl);

      const { getByText } = renderMain(managerUser, history);

      expect(history.location.pathname).toBe(newUserUrl);
      expect(getByText(/path does not exist!/i)).toBeInTheDocument();
    });

    it("should not redirect manager to user management page", async () => {
      const userUrl = routes.USERS;
      history.push(userUrl);

      const { getByText } = renderMain(managerUser, history);

      expect(history.location.pathname).toBe(userUrl);
      expect(getByText(/path does not exist!/i)).toBeInTheDocument();
    });
  });

  describe("Admin Navigation", () => {
    it("should redirect admin to new apartment page", () => {
      const newApartmentUrl = routes.NEW_APARTMENT;
      history.push(newApartmentUrl);

      const { getByText } = renderMain(adminUser, history);

      expect(history.location.pathname).toBe(newApartmentUrl);
      expect(getByText(/create new apartment/i)).toBeInTheDocument();
    });

    it("should redirect admin to new occupant page", async () => {
      const newOccupantUrl = routes.NEW_OCCUPANT;
      history.push(newOccupantUrl);

      const { getByText } = renderMain(adminUser, history);

      expect(history.location.pathname).toBe(newOccupantUrl);
      expect(getByText(/create new occupant/i)).toBeInTheDocument();
    });

    it("should redirect admin to new user page", async () => {
      const newUserUrl = routes.NEW_USER;
      history.push(newUserUrl);

      const { getByText } = renderMain(adminUser, history);

      expect(history.location.pathname).toBe(newUserUrl);
      expect(getByText(/create new user/i)).toBeInTheDocument();
    });

    it("should redirect admin to user management page", async () => {
      const userUrl = routes.USERS;
      history.push(userUrl);

      const { getByText } = renderMain(adminUser, history);

      expect(history.location.pathname).toBe(userUrl);
      expect(getByText(/role/i)).toBeInTheDocument();
    });
  });

  it("should show error message when landing on a bad page", () => {
    const history = createBrowserHistory();
    const badRoute = "/badRoute";
    history.push(badRoute);

    const { getByText } = renderMain(guestUser, history);

    expect(history.location.pathname).toBe(badRoute);
    expect(getByText("Path does not exist!")).toBeInTheDocument();
  });
});
