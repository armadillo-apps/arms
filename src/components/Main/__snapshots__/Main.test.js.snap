// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main renders correctly 1`] = `
<section
  className="app"
>
  <div
    className="containerCollapsed"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="linksContainer"
    >
      <a
        aria-current={null}
        className="links"
        href="/"
        onClick={[Function]}
      >
        A
      </a>
      <a
        aria-current="page"
        className="links active"
        data-testid="sideBar-apartments"
        href="/apartments"
        onClick={[Function]}
        style={Object {}}
      >
        <img
          alt="apartments"
          src="apartments.svg"
        />
        
      </a>
      <a
        aria-current={null}
        className="links"
        data-testid="sideBar-occupants"
        href="/occupants"
        onClick={[Function]}
      >
        <img
          alt="occupants"
          src="occupants.svg"
        />
        
      </a>
      <a
        aria-current={null}
        className="links"
        data-testid="sideBar-changepassword"
        href="/changePassword"
        onClick={[Function]}
      >
        <img
          alt="change password"
          src="edit.svg"
        />
        
      </a>
      
      <a
        aria-current="page"
        className="links active"
        href="/"
        onClick={[Function]}
        style={Object {}}
      >
        <img
          alt="logout"
          src="logout.svg"
        />
        
      </a>
    </div>
  </div>
  <div
    className="page"
    data-testid="apartments"
  >
    <div
      className="container"
    >
      <h1
        className="heading"
      >
        Apartments
      </h1>
      <div
        className="topBarContainer"
      >
        <input
          className="searchInput"
          data-testid="Apartment__searchBar"
          onChange={[Function]}
          placeholder="Search Apartment"
          type="text"
        />
        <div
          className="dateRangeContainer"
        >
          <div />
          <div
            className="DateRangePicker DateRangePicker_1"
          >
            <div>
              <div
                className="DateRangePickerInput DateRangePickerInput_1 DateRangePickerInput__showClearDates DateRangePickerInput__showClearDates_2"
              >
                <div
                  className="DateInput DateInput_1 DateInput__small DateInput__small_2"
                >
                  <input
                    aria-describedby="DateInput__screen-reader-message-startDateId"
                    aria-label="Start Date"
                    autoComplete="off"
                    className="DateInput_input DateInput_input_1 DateInput_input__small DateInput_input__small_2"
                    disabled={false}
                    id="startDateId"
                    name="startDateId"
                    onChange={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    placeholder="Start Date"
                    readOnly={false}
                    required={false}
                    type="text"
                    value=""
                  />
                  <p
                    className="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
                    id="DateInput__screen-reader-message-startDateId"
                  >
                    Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
                  </p>
                </div>
                <div
                  aria-hidden="true"
                  className="DateRangePickerInput_arrow DateRangePickerInput_arrow_1"
                  role="presentation"
                >
                  -
                </div>
                <div
                  className="DateInput DateInput_1 DateInput__small DateInput__small_2"
                >
                  <input
                    aria-describedby="DateInput__screen-reader-message-endDateId"
                    aria-label=" End Date"
                    autoComplete="off"
                    className="DateInput_input DateInput_input_1 DateInput_input__small DateInput_input__small_2"
                    disabled={false}
                    id="endDateId"
                    name="endDateId"
                    onChange={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    placeholder=" End Date"
                    readOnly={false}
                    required={false}
                    type="text"
                    value=""
                  />
                  <p
                    className="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
                    id="DateInput__screen-reader-message-endDateId"
                  >
                    Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
                  </p>
                </div>
                <button
                  aria-label="Clear Dates"
                  className="DateRangePickerInput_clearDates DateRangePickerInput_clearDates_1 DateRangePickerInput_clearDates__small DateRangePickerInput_clearDates__small_2 DateRangePickerInput_clearDates_default DateRangePickerInput_clearDates_default_3 DateRangePickerInput_clearDates__hide DateRangePickerInput_clearDates__hide_4"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  <svg
                    className="DateRangePickerInput_clearDates_svg DateRangePickerInput_clearDates_svg_1 DateRangePickerInput_clearDates_svg__small DateRangePickerInput_clearDates_svg__small_2"
                    focusable="false"
                    viewBox="0 0 12 12"
                  >
                    <path
                      d="M11.53.47a.75.75 0 0 0-1.061 0l-4.47 4.47L1.529.47A.75.75 0 1 0 .468 1.531l4.47 4.47-4.47 4.47a.75.75 0 1 0 1.061 1.061l4.47-4.47 4.47 4.47a.75.75 0 1 0 1.061-1.061l-4.47-4.47 4.47-4.47a.75.75 0 0 0 0-1.061z"
                      fillRule="evenodd"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <table
        className="table"
      >
        <thead
          className="tableHeader"
        >
          <tr>
            <th>
              Status
            </th>
            <th>
              Vacancy
            </th>
            <th>
              Apartment Name
            </th>
            <th>
              Lease Start
            </th>
            <th>
              Lease End
            </th>
            <th>
              Rental Per Month
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>
</section>
`;
